### PR TITLE
AVX-21420: add support for copilot sg mgmt

### PR DESCRIPTION
**Pull request by shuoliugit**
This is a feature for 6.8.

### CODE REVIEW

File: aviatrix/resource_aviatrix_copilot_security_group_management.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+	copilotSecurityGroupManagement := marshalCopilotSecurityGroupManagementInput(d)
+
+	if goaviatrix.IsCloudType(copilotSecurityGroupManagement.CloudType, goaviatrix.AWSRelatedCloudTypes|goaviatrix.AzureArmRelatedCloudTypes) &&
+		(copilotSecurityGroupManagement.Region == "" || copilotSecurityGroupManagement.Zone != "") {
+		return diag.Errorf("'region' is required and valid for AWS and Azure, 'zone' must be emtpy")
+	}
+
+	if goaviatrix.IsCloudType(copilotSecurityGroupManagement.CloudType, goaviatrix.GCPRelatedCloudTypes) &&
+		(copilotSecurityGroupManagement.Region != "" || copilotSecurityGroupManagement.Zone == "") {
+		return diag.Errorf("'zone' is required and valid for GCP, 'region' must be emtpy")
```
Comment (zhiqiangjin) [inline aviatrix/resource_aviatrix_copilot_security_group_management.go]: Please rephrase this error message.

File: aviatrix/resource_aviatrix_copilot_security_group_management.go [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+	}
+	return nil
+}
+
+func resourceAviatrixCopilotSecurityGroupManagementRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*goaviatrix.Client)
+
+	if d.Get("instance_id").(string) == "" {
+		id := d.Id()
+		log.Printf("[DEBUG] Looks like an import, no name received. Import Id is %s", id)
```
Comment (zhiqiangjin) [inline aviatrix/resource_aviatrix_copilot_security_group_management.go]: This error message is not right.

File: aviatrix/resource_aviatrix_copilot_security_group_management_test.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+}
+`, rName, os.Getenv("AWS_ACCOUNT_NUMBER"), os.Getenv("AWS_ACCESS_KEY"), os.Getenv("AWS_SECRET_KEY"),
+		os.Getenv("AWS_VPC_ID"), os.Getenv("AWS_REGION"), os.Getenv("AWS_SUBNET"))
+}
+
+func testAccCheckCopilotSecurityGroupManagementExists(resourceName string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		rs, ok := s.RootModule().Resources[resourceName]
+		if !ok {
+			return fmt.Errorf("copilot security group management not found: %s", resourceName)
```
Comment (zhiqiangjin) [inline aviatrix/resource_aviatrix_copilot_security_group_management_test.go]: Please work on this error message.

File: aviatrix/resource_aviatrix_copilot_security_group_management_test.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+		}
+
+		client := testAccProvider.Meta().(*goaviatrix.Client)
+
+		copilotSecurityGroupManagement, err := client.GetCopilotSecurityGroupManagement(context.Background())
+		if err != nil {
+			return err
+		}
+		if copilotSecurityGroupManagement.InstanceID != rs.Primary.ID {
+			return fmt.Errorf("could not find copilot security group management")
```
Comment (zhiqiangjin) [inline aviatrix/resource_aviatrix_copilot_security_group_management_test.go]: Please update this error message.

File: aviatrix/resource_aviatrix_copilot_security_group_management_test.go [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+	client := testAccProvider.Meta().(*goaviatrix.Client)
+
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "aviatrix_copilot_security_group_management" {
+			continue
+		}
+
+		_, err := client.GetCopilotSecurityGroupManagement(context.Background())
+		if err != goaviatrix.ErrNotFound {
+			return fmt.Errorf("copilot security group management still exists")
```
Comment (zhiqiangjin) [inline aviatrix/resource_aviatrix_copilot_security_group_management_test.go]: This error message is not right either.

**Review by zhiqiangjin — CHANGES_REQUESTED** [review]

File: aviatrix/resource_aviatrix_copilot_security_group_management_config.go [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+		if copilotSecurityGroupManagementConfig.CloudType != 0 || copilotSecurityGroupManagementConfig.AccountName != "" ||
+			copilotSecurityGroupManagementConfig.VpcId != "" || copilotSecurityGroupManagementConfig.InstanceId != "" ||
+			copilotSecurityGroupManagementConfig.Region != "" || copilotSecurityGroupManagementConfig.Zone != "" {
+			return diag.Errorf("'cloud_type', 'account_name', 'region'/'zone', 'vpc_id' and 'instance_id' are not needed to disable copilot security group management")
+		}
+
+		err := client.DisableCopilotSecurityGroupManagement(ctx)
+		if err != nil {
+			if strings.Contains(err.Error(), "not enabled") {
+				log.Printf("copilot security group management is already disabled")
```
Comment (mlin-aviatrix) [inline aviatrix/resource_aviatrix_copilot_security_group_management_config.go]: I feel like it shouldn't be an error to call the disable API when it is already disabled. For Private Mode, I check the error message in the `CheckFunc` and return nil if it is already disabled.

File: aviatrix/resource_aviatrix_copilot_security_group_management_config.go [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+	return resourceAviatrixCopilotSecurityGroupManagementConfigRead(ctx, d, meta)
+}
+
+func resourceAviatrixCopilotSecurityGroupManagementConfigDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*goaviatrix.Client)
+
+	err := client.DisableCopilotSecurityGroupManagement(ctx)
+	if err != nil {
+		if strings.Contains(err.Error(), "not enabled") {
+			log.Printf("copilot security group management is already disabled")
```
Comment (mlin-aviatrix) [inline aviatrix/resource_aviatrix_copilot_security_group_management_config.go]: Same.

File: goaviatrix/copilot_security_group_management_config.go [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+	LogEnable                            bool   `json:"log_enable,omitempty"`
+	State                                string `json:"state,omitempty"`
+}
+
+func (c *Client) EnableCopilotSecurityGroupManagement(ctx context.Context, copilotSecurityGroupManagementConfig *CopilotSecurityGroupManagementConfig) error {
+	copilotSecurityGroupManagementConfig.Action = "enable_copilot_sg"
+	copilotSecurityGroupManagementConfig.CID = c.CID
+	copilotSecurityGroupManagementConfig.LogEnable = true
+
+	err := c.PostAPIContext2(ctx, nil, copilotSecurityGroupManagementConfig.Action, copilotSecurityGroupManagementConfig, BasicCheck)
```
Comment (mlin-aviatrix) [inline goaviatrix/copilot_security_group_management_config.go]: Just `return c.PostAPIContext2(...)`

**Review by mlin-aviatrix — CHANGES_REQUESTED** [review]

**Review by mlin-aviatrix — APPROVED** [review]

**Review by zhiqiangjin — APPROVED** [review]

**Pull request merged**